### PR TITLE
VersionContext added

### PR DESCRIPTION
**Pull request by aslesarenko**

### CODE REVIEW

**Review by kushti — CHANGES_REQUESTED** [review]
Found some problems with updated ErgoLikeContext:

* remove default value for "val activatedScriptVersion: Byte'
* add the new field to JSON codecs

File: sigmastate/src/main/scala/sigmastate/serialization/SigmaSerializer.scala [hunk 1]
```diff
@@ -79,6 +93,7 @@ abstract class SigmaSerializer[TFamily, T <: TFamily] extends Serializer[TFamily
     val sigmaByteReader = new SigmaByteReader(r,
       new ConstantStore(),
       resolvePlaceholdersToConstants = false,
+      versionContext = VersionContext(0),
```
Comment (kushti) [inline sigmastate/src/main/scala/sigmastate/serialization/SigmaSerializer.scala:96]: Magic number

File: sigmastate/src/main/scala/sigmastate/serialization/SigmaSerializer.scala [hunk 2]
```diff
@@ -17,29 +18,42 @@ object SigmaSerializer {
   val MaxPropositionSize: Int = SigmaConstants.MaxPropositionBytes.value
   val MaxTreeDepth: Int = SigmaConstants.MaxTreeDepth.value
 
-    /** Helper function to be use in serializers.
+  //TODO v5.0: remove default value of versionContext parameter
+  // All usages of this method should pass a version context explicitly.
```
Comment (kushti) [inline sigmastate/src/main/scala/sigmastate/serialization/SigmaSerializer.scala:22]: Why not to do it now?

**Review by kushti — CHANGES_REQUESTED** [review]

File: sigmastate/src/main/scala/org/ergoplatform/JsonCodecs.scala [hunk 3]
```diff
@@ -400,7 +400,8 @@ trait JsonCodecs {
       "extension" -> ctx.extension.asJson,
       "validationSettings" -> ctx.validationSettings.asJson,
       "costLimit" -> ctx.costLimit.asJson,
-      "initCost" -> ctx.initCost.asJson
+      "initCost" -> ctx.initCost.asJson,
+      "activatedVersion" -> ctx.activatedScriptVersion.asJson
```
[the remark this answers is not in the document]
Comment (aslesarenko) [reply sigmastate/src/main/scala/org/ergoplatform/JsonCodecs.scala:404]: fixed

File: sigmastate/src/main/scala/org/ergoplatform/JsonCodecs.scala [hunk 4]
```diff
@@ -417,7 +418,11 @@ trait JsonCodecs {
       validationSettings <- cursor.downField("validationSettings").as[SigmaValidationSettings]
       costLimit <- cursor.downField("costLimit").as[Long]
       initCost <- cursor.downField("initCost").as[Long]
-    } yield new ErgoLikeContext(lastBlockUtxoRoot, Colls.fromArray(headers.toArray), preHeader,
-      dataBoxes, boxesToSpend, spendingTransaction, selfIndex, extension, validationSettings, costLimit, initCost)
+      version <- cursor.downField("activatedVersion").as[Byte]
```
[the remark this answers is not in the document]
Comment (aslesarenko) [reply sigmastate/src/main/scala/org/ergoplatform/JsonCodecs.scala:421]: fixed

File: sigmastate/src/main/scala/sigmastate/utils/SigmaByteReader.scala [hunk 5]
```diff
@@ -6,11 +6,28 @@ import sigmastate.Values.{SValue, Value}
...
+/** Reader used in the concrete implementations of [[SigmaSerializer]].
+  * It decorates the given reader, delegates most of the methods to it, but also adds new
+  * methods.
+  *
+  * @param r                              the underlying reader this reader reads from
+  * @param constantStore                  the store of constants which is used to resolve
+  *                                       [[sigmastate.Values.ConstantPlaceholder]]
+  * @param resolvePlaceholdersToConstants if true then resolved constants will be
+  *                                       substituted in the tree instead of the placeholder.
+  * @param versionContext                 optional version context to support soft and
```
[the remark this answers is not in the document]
Comment (aslesarenko) [reply sigmastate/src/main/scala/sigmastate/utils/SigmaByteReader.scala]: parameter removed

File: sigmastate/src/main/scala/sigmastate/Values.scala [hunk 6]
```diff
@@ -1058,6 +1058,13 @@ object Values {
       (header | version).toByte
     }
 
+    @inline def headerWithVersion(version: Byte): Byte = {
```
[the remark this answers is not in the document]
Comment (aslesarenko) [reply sigmastate/src/main/scala/sigmastate/Values.scala:1061]: added

File: sigmastate/src/main/scala/sigmastate/Values.scala [hunk 7]
```diff
@@ -1058,6 +1058,13 @@ object Values {
...
+      var h = updateVersionBits(DefaultHeader, version)
+      if (version > 0)
+        h = (h | ErgoTree.SizeFlag).toByte
```
[the remark this answers is not in the document]
Comment (aslesarenko) [reply sigmastate/src/main/scala/sigmastate/Values.scala:1064]: clarified

File: sigmastate/src/main/scala/sigmastate/interpreter/InterpreterContext.scala [hunk 8]
```diff
@@ -102,3 +104,15 @@ trait InterpreterContext {
...
+
+/** Descriptor of protocol versions.
+  * @param scriptVersion       the version of ErgoTree (see ErgoTree.header)
+  */
+case class VersionContext(scriptVersion: Byte)
+
+object VersionContext {
+
+  /** Descriptor of the max supported versions. */
+  val MaxSupportedVersion = VersionContext(
```
[the remark this answers is not in the document]
Comment (aslesarenko) [reply sigmastate/src/main/scala/sigmastate/interpreter/InterpreterContext.scala]: VersionContext removed entirely

File: sigmastate/src/main/scala/sigmastate/serialization/SigmaSerializer.scala (hunk 1 above)
Comment (aslesarenko) [reply sigmastate/src/main/scala/sigmastate/serialization/SigmaSerializer.scala]: removed

File: sigmastate/src/main/scala/sigmastate/serialization/SigmaSerializer.scala (hunk 2 above)
Comment (aslesarenko) [reply sigmastate/src/main/scala/sigmastate/serialization/SigmaSerializer.scala:21]: removed

**Review by kushti — APPROVED** [review]

**Pull request merged**